extensions/amp-3q-player/0.1/amp-3q-player.js

```diff
@@ -0,0 +1,190 @@
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {tryParseJson} from '../../../src/json';
+import {user} from '../../../src/log';
+import {removeElement} from '../../../src/dom';
+import {
+  installVideoManagerForDoc,
+} from '../../../src/service/video-manager-impl';
+import {isObject} from '../../../src/types';
+import {VideoEvents} from '../../../src/video-interface';
+import {videoManagerForDoc} from '../../../src/video-manager';
+
+/**
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+class Amp3QPlayer extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.playerReadyPromise_ = null;
+
+    /** @private {?Function} */
+    this.playerReadyResolver_ = null;
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url('https://playout.3qsdn.com', opt_onLayout);
+  }
+
+  /** @override */
+  buildCallback() {
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    this.iframe_ = iframe;
+
+    this.forwardEvents([VideoEvents.PLAY, VideoEvents.PAUSE], iframe);
+    this.applyFillContent(iframe, true);
+    this.element.appendChild(iframe);
+
+    iframe.setAttribute('frameborder', '0');
+    iframe.setAttribute('allowfullscreen', 'true');
+
+    installVideoManagerForDoc(this.element);
+    videoManagerForDoc(this.element).register(this);
+  }
+
+  /** @override */
+  layoutCallback() {
+
+    const dataId = user().assert(
+        this.element.getAttribute('data-id'),
+        'The data-id attribute is required for <amp-3q-player> %s',
+        this.element);
+
+    const src = 'https://playout.3qsdn.com/' + dataId + '?autoplay=false&amp=true';
+    this.iframe_.src = src;
+
+    window.addEventListener('message',
+                            event => this.sdnBridge_(event));
+
+    return this.loadPromise(this.iframe_)
+      .then(() => {
+        this.element.dispatchCustomEvent(VideoEvents.LOAD);
+        this.playerReadyResolver_(this.iframe_);
+      });
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {visible});
+  }
+
+  /** @override */
+  pauseCallback() {
+    if (this.iframe_) {
+      this.pause();
+    }
+  }
+
+  /** @private */
+  sdnBridge_(event) {
+    const data = isObject(event.data) ? event.data : tryParseJson(event.data);
```

Comment:
This does not seem to work based on my testing. Please see https://amphtml-ae5fc.firebaseapp.com/3q/examples/empty.amp.html and open the DevTools console and interact with the platey. The only message I can see is:
```
ima://{"name":"gpt","type":"isGptPresent","data":{"scope":"proxy"},"sid":"*","channel":"goog_569777204"}
```

nothing with play, pause, mute, unmute events.